classes/DirtEventTypes.sc

```diff
@@ -36,91 +36,94 @@ DirtEventTypes {
 
 
 	midiEvent = (
+			midiOutNotFoundError: DirtPartTimeError(2, "midi device is nil."),
 			play: #{
 
-				var freq, lag, sustain, func;
-				var args, midiout, hasGate, midicmd, latency;
-				var sendNRPN;
+				var freq, lag, sustain;
+				var args, midiout, hasGate, midicmd, latency, chan;
+				var sendNRPN, schedmidi, schedmidicmd;
+				var hasNote = ~n != \none;
 
-				midicmd = ~midicmd;
 				midiout = ~midiout.value;
 
-				if(midicmd.isNil) {
-					if(~ccn.notNil) { midicmd = \control; ~ctlNum = ~ccn };
-					if(~ccv.notNil) { midicmd = \control; ~control = ~ccv };
-					if(~nrpn.notNil) {
-						var chan, nrpnMSB, nrpnLSB, valMSB, valLSB;
-						~val = ~val ? 0;
-						chan = ~midichan ? 0;
-						midicmd = \control;
-						nrpnLSB = ~nrpn % 128;
-						nrpnMSB = (~nrpn - nrpnLSB) / 128;
-						valLSB  = ~val % 128;
-						valMSB  = ((~val - valLSB) / 128) % 128;
-						sendNRPN = {
-							midiout.control(chan, 99, nrpnMSB);
-							midiout.control(chan, 98, nrpnLSB);
-							midiout.control(chan, 6,  valMSB);
-							midiout.control(chan, 38, valLSB)
-						};
-					};
-					if(~progNum.notNil) { midicmd = \program };
-					if(~polyTouch.notNil) { midicmd = \polyTouch };
-					if(~midibend.notNil) { midicmd = \bend; ~val = ~midibend; };
-					if(~miditouch.notNil) { midicmd = \touch; ~val = ~miditouch; };
-					if(midicmd.isNil) { midicmd = \noteOn }; // if still nil
+				if(midiout.isNil) {
+					~midiOutNotFoundError.throw;
 				};
+				midicmd = ~midicmd;
+				~chan = ~midichan ? 0;
+				chan = ~chan;
 
-
-				freq = ~freq.value;
-
-				~amp = ~amp.value;
-				~midinote = (freq.cpsmidi).round(1).asInteger;
 				lag = ~lag + (~latency ? 0);
-				sustain = ~sustain = ~sustain.value;
-				if(~uid.notNil and: { midiout.notNil }) {
-					~uid = midiout.uid    // mainly for sysex cmd
+				latency = lag; // for now.
+				schedmidi = if (latency == 0.0) {
+					{|f| f.value}
+				} {
+					{|f| thisThread.clock.sched(latency, f)}
+				};
+				schedmidicmd = { |cmd|
+					var func;
+					func = Event.default[\midiEventFunctions][cmd];
+					args = func.valueEnvir.asCollection;
+					schedmidi.value { midiout.performList(cmd, args) };
+					if (midicmd.notNil && midicmd === cmd) {
```

Comment:
this should be: `midicmd.notNil && (midicmd === cmd)` sclang always binds to the left. More efficient in this case is: `midicmd.notNil and: { midicmd === cmd }` which inlines everything so you have fewer function calls.

Reply:
Ok thanks, yes this bug resulted in doubling up of events. I created a new PR for this.